tools/cmaketobzl/cmaketobzl.go

```diff
@@ -0,0 +1,389 @@
+	if text == bc.begin {
+		bc.count += 1
+	} else if text == bc.end {
+		bc.count -= 1
+	} else {
+		matched = false
+	}
+	return matched || bc.count > 0
+}
+
+type eval struct {
+	p *ast.Parser
+	o options
+
+	w    *writer.StarlarkWriter
+	v    *bindings.Mapping
+	path []string
+}
+
+type options struct {
+	macroName   string
+	shouldPrint func(string) bool
+	shouldAdd   func(string) bool
+	excludePath func(string) bool
+}
+
+// Option is a configuration option for the CMake evaluator.
+type Option func(*eval)
+
+// PrintCommands configures the evaluator to print commands on the StarlarkWriter for which the supplied predicate returns true.
+func PrintCommands(p func(string) bool) Option {
+	return func(e *eval) {
+		e.o.shouldPrint = p
+	}
+}
+
+// RecurseCommands configures the evaluator to recurse into the subdirectory
+// specified by the first argument to the command when the provided predicate returns true.
+// By default only "add_subdirectory" is handled this way.
+func RecurseCommands(p func(string) bool) Option {
+	return func(e *eval) {
+		e.o.shouldAdd = p
+	}
+}
+
+// ExcludePaths configures the evaluator to omit particular paths entirely during traversal.
+func ExcludePaths(p func(string) bool) Option {
+	return func(e *eval) {
+		e.o.excludePath = p
+	}
+}
+
+// DefineVars configures the evaluator to predefine the specified variables.
+func DefineVars(vars map[string]string) Option {
+	return func(e *eval) {
+		for k, v := range vars {
+			e.v.Set(k, v)
+		}
+	}
+}
+
+// Matching compiles the provided pattern and returns a predicate for matching strings.
+func Matching(pat string) func(string) bool {
+	return regexp.MustCompile(pat).MatchString
+}
+
+// NewEvaluator returns a new CMake evaluator instance.
+func NewEvaluator(w io.Writer, opts ...Option) *eval {
+	e := &eval{
+		p: ast.NewParser(),
+		w: writer.NewStarlarkWriter(w),
+		v: bindings.New(),
+		o: options{
+			macroName: "generated_cmake_targets",
+			shouldAdd: func(n string) bool { return n == "add_subdirectory" },
+		},
+	}
+	for _, o := range opts {
+		o(e)
+	}
+	return e
+}
+
+// parse parses the provided input into a CMakeFile AST.
+func (e *eval) parse(input io.Reader) (*ast.CMakeFile, error) {
+	return e.p.Parse(input)
+}
+
+// parse parses the provided path into a CMakeFile AST.
+func (e *eval) parseFile(path string) (*ast.CMakeFile, error) {
+	input, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer input.Close()
+	return e.parse(input)
+}
+
+// walk evaluates all of the provided CMakeLists.txt files into the body of a single Starlark macro..
+func (e *eval) walk(paths []string) error {
+	if err := e.w.BeginMacro(e.o.macroName); err != nil {
+		return err
+	}
+	root, paths := SplitCommonRoot(paths)
+	e.path = append(e.path, root)
+	for _, p := range paths {
+		if err := e.AddSubdirectory(p); err != nil {
+			return err
+		}
+	}
+	return e.w.EndMacro()
+}
+
+// dispatchFunc is a function which handles the current command, updates the
```

Comment:
The name `dispatchFunc` suggests to me that it dispatches something—whereas, I think, it's really more of a "handler" for a particular command, and is _returned_ by the `dispatch` method to handle the command.

Reply:
Is fair, the model and 'Fn' naming were borrowed from https://golang.org/src/text/template/parse/lex.go

dispatch itself is a valid (and the default) handler, though (and generally returns itself).  They aren't handlers for the specific command in question but are intended to allow executing if, foreach, etc. while retaining the current state so they need to control dispatch of commands which occur between the opening and closing command.